tor: enable lnd surviving Tor daemon restart

yyforyongyu opened the pull request:
This PR made two changes to the tor controller,
- a new health check, tor connection, upon startup. It will detect the current hidden service status and re-connect the Tor daemon once it finishes restarting.
- during `lnd` shutdown, the onion service is removed.

To accomplish the task, a new response parser is added to the tor controller, along with more unit tests.

Fix #5166

Roasbeef commented on tor/cmd_info.go:
```diff
@@ -0,0 +1,65 @@
+	// single source to change our state.
+	case success:
+		log.Errorf("server replied 250, but with an err:%v", err)
+		return nil
```
Yet we don't do anything with this? Or this is just mean to allow us to check if the tor daemon knows of the service or not?

yyforyongyu replied:
Ok this usage of the endpoint seems to have created some confusion. I've switched to a different endpoint, the `GETINFO onions/current`, which would also save us a few bandwidths as the rely is much smaller.

Roasbeef commented on tor/controller.go:
```diff
@@ -169,6 +177,56 @@ func (c *Controller) Stop() error {
+	}
+
+	// Exit early if we have no activeServiceID set.
+	if c.activeServiceID == "" {
```
Which would mean that auth above failed?

Roasbeef commented on healthcheck/tor_connection.go:
```diff
@@ -0,0 +1,85 @@
+// tor controller. If the tor daemon comes back, a new socket connection will
+// then be created.
+func CheckTorServiceStatus(tc *tor.Controller, createService func() error) error {
+	err := tc.GetServiceInfo()
```
Ah ok question answered re prior comment, maybe change the name to reflect this is itself more of a health check and we don't actually parse the response?

yyforyongyu replied:
The function is now renamed to `CheckOnionService`

Roasbeef commented on healthcheck/tor_connection.go:
```diff
@@ -0,0 +1,85 @@
+		return fmt.Errorf("Tor daemon connection lost, " +
+			"check if Tor is up and running")
+
+	// We will get a broken pipe error when we use the old connection to
```
Old connection here is the existing connection we have before the tor daemon restarted (in this scenario)?

yyforyongyu replied:
Yes! Revised the docs to make it more clear.

Roasbeef commented on server.go:
```diff
@@ -1362,9 +1362,40 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+	return s, nil
+}
+
+// createLivenessMonitor creates a set of health checks using our configured
```
👍  for spinning this out into a new connection

carlaKC commented on tor/cmd_onion.go:
```diff
@@ -254,3 +257,38 @@ func (c *Controller) AddOnion(cfg AddOnionConfig) (*OnionAddr, error) {
+// previously created via an "ADD_ONION" command. It is only possible to remove
+// Onion Services that were created on the same control connection as the
+// "DEL_ONION" command, and those that belong to no control connection in
+// particular (The "Detach" flag was specified at creation).
```
nit: clarify this "and those that belong to no control connection in particular if the "Detach" flag ..."

yyforyongyu replied:
Updated!

carlaKC commented on tor/cmd_onion.go:
```diff
@@ -254,3 +257,38 @@ func (c *Controller) AddOnion(cfg AddOnionConfig) (*OnionAddr, error) {
+	// await its response.
+	code, _, err := c.sendCommand(cmd)
+
+	// Tor replies with "250 OK" on success, or a 512 if there are an
```
Say you're running lnd and then upgrade to this change, is it possible that we try to `DelOnion` on an onion service that does not belong to a control connection and doesn't have "Detatch" set? (ie, was created before this PR)

Wondering what error code that returns/ whether we need to handle it?

yyforyongyu replied:
Yeah it was a bit confusing, I've updated the docs on `AddOnion`, hopefully that would make things a bit clearer.

For the specific case, no it won't happen as,
- when running old lnd, the old Onion Service would belong to the old control connection, so it cannot be accessed by other connections.
- when upgrading, lnd needs to be restarted, thus the control connection will be re-created, and the old Onion Service would be gone, a new one will be created.

From the specs, there are three possible response codes we can get, 250 on success, 512 on invalid arguments, and 552 if the `ServiceID` is not recognized. I've updated the docs below to make it more verbose on code 552.

carlaKC commented on tor/cmd_info.go:
```diff
@@ -0,0 +1,65 @@
+// unrecognized or not found in its local descriptor cache.
+func (c *Controller) GetServiceInfo() error {
+	// Check that we have a hidden service created.
+	if c.activeServiceID == "" {
```
We ok for read/write races here? Seems ok so far, just wondering.

carlaKC commented on tor/controller.go:
```diff
@@ -206,9 +206,11 @@ func (c *Controller) sendCommand(command string) (int, string, error) {
-// Unless specified otherwise, multiple lines in a single reply from Tor to the
-// controller are guaranteed to share the same status code. Read more on this
-// topic: https://gitweb.torproject.org/torspec.git/tree/control-spec.txt#n158
+// Unless specified otherwise, multiple lines in a single reply from Tor daemon
```
nit: move this to previous commit?

yyforyongyu replied:
Done!

carlaKC commented on tor/controller.go:
```diff
@@ -169,6 +177,56 @@ func (c *Controller) Stop() error {
+
+	// Remove the old Onion Service. The service might not be found once
+	// the Tor daemon restarted, so we ignore the error here.
+	if err := c.DelOnion(c.activeServiceID); err != nil {
```
nit: add a comment here that `activeServiceID` is still the old service? Had to go read `authenticate` to check that it doesn't update this value + above maybe clarify that it would only be set for old services?

yyforyongyu replied:
Updated!

carlaKC commented on healthcheck/tor_connection.go:
```diff
@@ -0,0 +1,85 @@
+	// return an error and wait for the Tor daemon to come back. We won't
+	// attempt to make a new connection since we know Tor daemon is down.
+	case errors.Is(err, io.EOF), errors.Is(err, syscall.ECONNREFUSED):
+		log.Warnf("Tor connection lost, attempting a tor controller " +
```
This log is a bit unintuitive, we might not actually attempt re-connection in the next check if tor continues to give us an EOF? Also thing that the healthcheck will log the error we return so we don't need to log and return here?

yyforyongyu replied:
I think we would only ever get one EOF, which signals the connection is lost. And yeah, if it keeps giving EOF, there's no point to re-connect because there's no socket available.

The warning log is removed now.

yyforyongyu replied:
A second thought. I think the healthcheck should give warning rather than info so the user can act upon it?

carlaKC reviewed (commented):
Fantastic thorough testing in this PR, love to see it! 

Finished a somewhat high-level pass, will come back after re-reading tor docs + testing locally and give this another look.

yyforyongyu commented on tor/log.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+// Subsystem defines the logging code for this subsystem.
+const Subsystem = "TCTR" // TCTR as in Tor Controller.
```
Changed.

yyforyongyu commented on tor/cmd_onion.go:
```diff
@@ -254,3 +257,38 @@ func (c *Controller) AddOnion(cfg AddOnionConfig) (*OnionAddr, error) {
+// DelOnion tells the server to remove an Onion ("Hidden") Service, that was
+// previously created via an "ADD_ONION" command. It is only possible to remove
+// Onion Services that were created on the same control connection as the
+// "DEL_ONION" command, and those that belong to no control connection in
```
Updated the docs to make it more clear.

yyforyongyu commented on healthcheck/tor_connection.go:
```diff
@@ -0,0 +1,85 @@
+			"check if Tor is up and running")
+
+	// We will get a broken pipe error when we use the old connection to
+	// make a get info request. The Tor daemon might yet be running. In
```
Thanks for the review, now fixed!

yyforyongyu commented on healthcheck/tor_connection.go:
```diff
@@ -0,0 +1,85 @@
+	// ErrServiceNotCreated or ErrServiceNotFound, there's little we can do
+	// but to report the error to the user.
+	default:
+		return err // NOTE: err could be nil.
```
yeah right, comments updated.

yyforyongyu commented:
This is new,
```go
--- FAIL: TestShutdownIfChannelClean (0.16s)
    link_test.go:6558: 
        	Error Trace:	link_test.go:6558
        	            				link_test.go:6609
        	Error:      	An error is expected but got nil.
        	Test:       	TestShutdownIfChannelClean
        	Messages:   	link failed to shutdown
FAIL
coverage: 21.6
FAIL	github.com/lightningnetwork/lnd/htlcswitch	134.820s
```

yyforyongyu commented:
Major change from last commit,
- removed `Detach` flag
- use `GETINFO onions/current` instead of `GETINFO hs/service/desc/id/<ADDR>`
- renamed methods related to healthcheck

PTAL @carlaKC @Roasbeef

Roasbeef commented:
> This is new,



Should be fixed by https://github.com/lightningnetwork/lnd/pull/5808
